elasticsearch/src/main/java/com/tcdi/zombodb/query/VisibilityQueryHelper.java

```diff
@@ -18,167 +18,260 @@
 import org.apache.lucene.index.*;
 import org.apache.lucene.queries.TermFilter;
 import org.apache.lucene.queries.TermsFilter;
-import org.apache.lucene.search.FieldCache;
+import org.apache.lucene.search.Filter;
 import org.apache.lucene.search.IndexSearcher;
-import org.apache.lucene.search.MatchAllDocsQuery;
-import org.apache.lucene.search.Query;
+import org.apache.lucene.search.NumericRangeFilter;
 import org.apache.lucene.search.join.ZomboDBTermsCollector;
-import org.apache.lucene.util.*;
+import org.apache.lucene.store.ByteArrayDataInput;
+import org.apache.lucene.util.BytesRef;
+import org.apache.lucene.util.BytesRefBuilder;
+import org.apache.lucene.util.FixedBitSet;
+import org.apache.lucene.util.NumericUtils;
+import org.elasticsearch.common.hppc.IntOpenHashSet;
+import org.elasticsearch.common.hppc.IntSet;
+import org.elasticsearch.common.hppc.cursors.IntCursor;
+import org.elasticsearch.common.lucene.search.AndFilter;
+import org.elasticsearch.common.lucene.search.MatchAllDocsFilter;
+import org.elasticsearch.common.lucene.search.OrFilter;
 import org.elasticsearch.common.lucene.search.XConstantScoreQuery;
-import org.elasticsearch.search.internal.SearchContext;
 
 import java.io.IOException;
 import java.util.*;
-import java.util.concurrent.ConcurrentSkipListSet;
 
 final class VisibilityQueryHelper {
 
-    private static final ConcurrentSkipListSet<Long> KNOWN_COMMITTED_XIDS = new ConcurrentSkipListSet<>();
+    private static class HeapTuple implements Comparable<HeapTuple> {
```

Comment:
Pull this out into a standalone class and put some unit tests around its `.equals()` and `.compareTo()` methods as they were a source of bugs during development